services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java

```diff
@@ -0,0 +1,296 @@
+                            return signal.implementsSchemaVersion(schemaVersion) ? signal
+                                    : CommandNotSupportedException.newBuilder(schemaVersion.toInt())
+                                    .dittoHeaders(dittoHeaders)
+                                    .build();
+                        } catch (final Exception e) {
+                            return new Status.Failure(e);
+                        }
+                    })
+                    .to(Sink.actorRef(createHttpPerRequestActor(ctx, httpResponseFuture),
+                            AbstractHttpRequestActor.COMPLETE_MESSAGE))
+            );
+
+            // return with future
+
+            return completeWithFuture(httpResponseFuture);
+
+        });
+
+    }
+
+    /**
+     * Convert the request and payload to a cloud event.
+     *
+     * @param ctx The request context.
+     * @param payload The binary payload, this may contain the cloud event payload, or the fully encoded cloud
+     *         event structure.
+     * @return The cloud event
+     */
+    private CloudEvent toCloudEvent(final RequestContext ctx, final DittoHeaders dittoHeaders, final ByteString payload) {
+
+        if (LOGGER.isTraceEnabled()) {
+
+            final StringBuilder headers = new StringBuilder("Raw HTTP Headers:");
+            ctx.getRequest().getHeaders()
+                    .forEach(header -> headers
+                            .append("\n\t")
+                            .append(header.name())
+                            .append(" = ")
+                            .append(header.value()));
+
+            LOGGER
+                    .withCorrelationId(dittoHeaders)
+                    .trace(headers.toString());
+
+            LOGGER
+                    .withCorrelationId(dittoHeaders)
+                    .trace("Ditto: {}", dittoHeaders);
+        }
+
+        // create a reader for the message
+        final MessageReader reader = HttpMessageFactory.createReader(acceptor -> {
+
+            // NOTE: this acceptor may be run multiple times by the message reader
+
+            // record if we saw the content type header
+            final AtomicBoolean sawContentType = new AtomicBoolean();
+            // consume the HTTP request headers
+            ctx.getRequest().getHeaders().forEach(header -> {
+                if (header.lowercaseName().equals(DittoHeaderDefinition.CONTENT_TYPE.getKey())) {
+                    sawContentType.set(true);
+                }
+                acceptor.accept(header.name(), header.value());
+            });
+
+            if (!sawContentType.get()) {
+                // we didn't see the content type in the header, so extract it from akka's request
+                acceptor.accept(DittoHeaderDefinition.CONTENT_TYPE.getKey(), ctx.getRequest().entity().getContentType().mediaType().toString());
+            }
+        }, payload.toArray());
+
+        return reader.toEvent();
```

Comment:
I just tried out invoking the Route manually.
This definitely needs to catch `CloudEventRWException` and `IllegalStateException`. `IllegalStateException` is for example thrown if invoked without "cloudevent" headers.
That currently causes an 500 "internal server" error because of the unhandled exception.

I suggest catching the exceptions here and throwing a new DittoRuntimeException instead, e.g. `CloudEventNotParsableException` with status code 400 "bad request" giving a hint that a HTTP cloudevent was expected and e.g. pointing to the spec in the "href" of the exception.

Reply:
I updated the PR, checking and handling the error. Please take a look, I am not sure I was following "best practice" of Ditto error handling.

Reply:
Cool, looks good 👍